applications/DeepSpeed-Chat/training/step3_rlhf_finetuning/ppo_trainer.py

```diff
@@ -71,8 +71,7 @@ def _generate_sequence(self, prompts):
 
         with torch.no_grad():
             seq = self.actor_model.module.generate(prompts,
-                                                   max_length=max_min_length,
-                                                   min_length=max_min_length)
+                                                   max_length=max_min_length)
```

Comment:
The min_length is added for our Hybrid Engine.

Reply:
I am not sure if my understanding is correct or not. Although, in the original code, the max length and min_length is set to the same number max_min_length, which will force the model generate to the max length, which leads to repeated generation. Last time I modify the code to run my own model and dataset then I occurs the repeated generation. I will try to run the original code to see whether this is still a problem.

Reply:
@yaozhewei  Thanks for you reply! When I tried the original code, I found there are several detail problems which might be issue for the training.

Reply:
@yaozhewei Thanks for you reply! When I tried the original code, I found there are several detailed problems which might be issues for the training. 

1. **The different tokenizer performance for tokenizer before and after save vocabulary.**  
 <img width="885" alt="image" src="https://user-images.githubusercontent.com/117286828/234270818-c729507a-53b8-441e-80ca-b10cdae27aee.png"> As shown in the image, the tokenizer load from facebook/opt-1.3b will add extra pad token '\</s>' as the prefix and it treats '<|endoftext|>' as different tokens. Although the tokenizer load from the actor model path(sft model saved in stage1) does not add extra </s> and treats '<|endoftext|>' as a single token. It might causes the inconsistent between stage1 and stage3.
2. **The padding issue causes the wrong generation.The default padding side for OPT model is right. We padding the prompt to max max_prompt_seq_len in stage3, which will make the model generate nonsense.**
<img width="1367" alt="image" src="https://user-images.githubusercontent.com/117286828/234274504-583c7217-4b29-4d3d-ac1a-26ecba7e01b2.png">
As shown in the image, the DataCollatorRLHF will pad the prompt to max prompt length which is 256 by default. The first cell which include padding will keep generate nonsense. (In some case it generates some incorrect response, such as starting by 'Human: '. The second without padding performances correctly. One possible solution will be setting the tokenizer.padding_side='left'.

3. **The min_length setting force the model generate to max length, which produce repeated or nonsense result.** 
<img width="1364" alt="image" src="https://user-images.githubusercontent.com/117286828/234278112-332db8fd-ffaa-4622-a1fa-2d7d521ff630.png">
Pick up the same sample as the example. We can see the last generation result contains nonsense in the end of generation result.


I think these issues will affect the training of RLHF because the generated experience are all bad.

Reply:
The links in your second bullet point are wrong, could you update it? thanks

Reply:
More update for 1:
It probably an issue from HF (origianlly, I misunderstood the problem. The <|endoftext|> token shouldn't be modified as you pointed out). We tried the following 
```
import transformers 
from transformers import AutoModelForCausalLM, AutoTokenizer, AutoConfig
import os 
import torch 



pretrained = "facebook/opt-125m" # "bigscience/bloom-560m"
tokenizer = AutoTokenizer.from_pretrained(pretrained,
                                              fast_tokenizer=True)
model_config = AutoConfig.from_pretrained(pretrained)
model = AutoModelForCausalLM.from_pretrained(
        pretrained,
        from_tf=bool(".ckpt" in pretrained),
        config=model_config)

print("original: ")
print(tokenizer.pad_token, tokenizer.bos_token, tokenizer.eos_token)

output_dir = "./opt/"
model_to_save = model.module if hasattr(model, 'module') else model
CONFIG_NAME = "config.json"
WEIGHTS_NAME = "pytorch_model.bin"
if not os.path.exists(output_dir):
    os.makedirs(output_dir)
output_model_file = os.path.join(output_dir, WEIGHTS_NAME)
output_config_file = os.path.join(output_dir, CONFIG_NAME)
save_dict = model_to_save.state_dict()
torch.save(save_dict, output_model_file)
model_to_save.config.to_json_file(output_config_file)
tokenizer.save_vocabulary(output_dir)

tokenizer_save = AutoTokenizer.from_pretrained(output_dir,
                                                fast_tokenizer=True)
print("after save: ")
print(tokenizer_save.pad_token, tokenizer_save.bos_token, tokenizer_save.eos_token)
```

and here is what we got:
```
original:
<pad> </s> </s>
after save:
Using pad_token, but it is not set yet.
None <|endoftext|> <|endoftext|>
```

This is related to https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/tokenization_gpt2_fast.py#L129-L139. The special token treatment is hard coded and it won't read the config.json to know how to deal with special tokens. We will provide a workaround soon.

Reply:
Could you let me know how you enable left padding in the repo? You mentioned flip the sentence before padding and then flip it back. The links in your second bullet point are wrong, could you update it? thanks

Reply:
@DanqingZ updated. I have created the PR to resolve issue 1: https://github.com/microsoft/DeepSpeedExamples/pull/460